zerver/tests/test_auth_backends.py

```diff
@@ -1054,26 +1062,56 @@ def stage_two_of_registration(self, result: HttpResponse, realm: Realm, subdomai
         else:
             self.assertIn('do_confirm/' + confirmation_key, result.url)
             do_confirm_url = result.url
+
+        if user_avatar_url:
+            # If there is an avatar_url, then there must be a user_avatar_url
+            # on the preregistration user.
+            prereg_user = confirmation.content_object
+            self.assertEqual(prereg_user.user_avatar_url, user_avatar_url)
+
         result = self.client_get(do_confirm_url, name = name)
         self.assert_in_response('action="/accounts/register/"', result)
         data = {"from_confirmation": "1",
                 "key": confirmation_key}
+
         result = self.client_post('/accounts/register/', data)
         if not skip_registration_form:
             self.assert_in_response("We just need you to do one last thing", result)
 
             # Verify that the user is asked for name but not password
             self.assert_not_in_success_response(['id_password'], result)
             self.assert_in_success_response(['id_full_name'], result)
+            if user_avatar_url:
+                # Verify if there is a checkbox for the user choose to use their
+                # social avatar.
+                self.assert_in_success_response(['use_social_avatar'], result)
             # Verify the name field gets correctly pre-populated:
             self.assert_in_success_response([expected_final_name], result)
 
-            # Click confirm registration button.
-            result = self.client_post(
-                '/accounts/register/',
-                {'full_name': expected_final_name,
-                 'key': confirmation_key,
-                 'terms': True})
+            # When a user choose not to use their social avatar, the registered url won't be accessed.
+            with responses.RequestsMock(assert_all_requests_are_fired=False) as mock_upload:
+                if mocked_social_avatar:
+                    mock_upload.add(
+                        method='GET',
+                        url=user_avatar_url,
+                        status=200,
+                        content_type='binary',
+                        body=mocked_social_avatar.getvalue()
+                    )
+
+                post_dict = {'full_name': expected_final_name,
+                             'key': confirmation_key,
+                             'terms': True}
+                post_dict['use_social_avatar'] = 'on' if use_social_avatar else ''
+                result = self.client_post('/accounts/register/', post_dict)
+
+                user = get_user_by_delivery_email(email, realm)
+                response_social_avatar_hash = ''
+                if use_social_avatar:
+                    response_social_avatar_hash = user_avatar_content_hash(mock_upload.calls[0].response.content)
+
+                self.verify_user_avatar(user, user_avatar_url, use_social_avatar, mocked_social_avatar,
+                                        response_social_avatar_hash)
```

Comment:
So here the auth flow has ended and we start doing all the asserts to verify the outcome. I don't know if it's good for the avatar verification to be the first thing checked. Imagine a test unrelated to avatars  is failing in the future due to some bugs in the "core' logic, not something specific to avatars. With this order of verification of things, the developer running the test will see it failing in `verify_user_avatar`, which is likely to be confusing. This is why I think the avatar verification should happen in the code blocks below this, after the more basic correctness checks.